imagebuildah/stage_executor.go

```diff
@@ -1174,23 +1197,17 @@ func (s *StageExecutor) Execute(ctx context.Context, base string) (imgID string,
 		// to find the digest of the content to check for a cached
 		// image, run the step so that we can check if the result
 		// matches a cache.
-		if cacheID == "" {
+		// We already called ib.Run() for the `canMatchCacheOnlyAfterRun`
+		// cases above, so we shouldn't do it again.
+		if cacheID == "" && !canMatchCacheOnlyAfterRun {
 			// Process the instruction directly.
 			if err = ib.Run(step, s, noRunsRemaining); err != nil {
 				logrus.Debugf("Error building at step %+v: %v", *step, err)
 				return "", nil, fmt.Errorf("error building at STEP \"%s\": %w", step.Message, err)
 			}
 
 			// In case we added content, retrieve its digest.
-			addedContentType, addedContentDigest := s.builder.ContentDigester.Digest()
-			addedContentSummary = addedContentType
-			if addedContentDigest != "" {
-				if addedContentSummary != "" {
-					addedContentSummary = addedContentSummary + ":"
-				}
-				addedContentSummary = addedContentSummary + addedContentDigest.Encoded()
-				logrus.Debugf("added content %s", addedContentSummary)
-			}
+			addedContentSummary = s.getContentSummaryAfterAddingContent()
```

Comment:
Why are we searching the cache again? In the `!canMatchCacheOnlyAfterRun` case, we should have had the right `addedContentSummary` initially already; so this shouldn’t find anything new.[1]

Also, the “In case we added content, retrieve its digest” comment is now misleading, that shouldn’t happen / make a difference on this code path.

[1] Unless, I guess, some parallel build has created an equivalent image in the meantime. But that’s _always_ a race.